### PR TITLE
[DOC/ENH] MWE and fixes for `truncate_datetime`

### PR DESCRIPTION
**Pull request by thatlittleboy**
# PR Description

Please describe the changes proposed in the pull request:
- add MWE.
- make `truncate_datetime_dataframe` not mutate original df.
- fix SettingWithCopyWarning; and typo bug with the "MINUTE:" key in `ENUM` dict of
    `_truncate_datetime` function.
- ditch the previous error-catching mechanism. Optimize to early exit if no datetime columns found in df; Only apply `_truncate_datetime` on datetime-like columns, and when NaT is encountered, it is ignored.
- small optimization to exit immediately if wrong datepart is passed in, instead of carrying on iterating
    through all elements.
- not iterating over rows, and instead using `.applymap` on relevant columns only yields about 60x improvement. (5+ seconds -> 80ms on a df with 10k timestamps).
- fix warning about use of `pd.np` in tests for `to_datetime`.
- make tests more complete (test all dateparts instead of just MONTH, and test specifically when there is `pd.NaT`)
- administrative: collate the tests into just 1 file for this `truncate_datetime` function.

**This PR is related to #972**

### CODE REVIEW

**Comment by ericmjl** [comment]
🚀 Deployed on https://deploy-preview-1040--pyjanitor.netlify.app

**Review by samukweku — APPROVED** [review]

File: janitor/functions/truncate_datetime.py [hunk 1]
```diff
@@ -15,46 +17,79 @@ def truncate_datetime_dataframe(
...
+    datepart = datepart.upper()
+    if datepart not in ACCEPTABLE_DATEPARTS:
+        raise ValueError(
+            "Received an invalid `datepart` precision. "
+            f"Please enter any one of {ACCEPTABLE_DATEPARTS}."
+        )
+
+    df = df.copy()
+
+    for col in df.columns:
```
Comment (samukweku) [inline janitor/functions/truncate_datetime.py]: Apologies, I know I approved ... Just couldn't hake this off, what is the purpose of iterating through the index? Isn't the same function applied to every row?

Comment (thatlittleboy) [reply janitor/functions/truncate_datetime.py]: No worries,
Yeah, it is. Are you implying we can use `.apply` here?

(I originally started this PR only with the intention to add a MWE, but then realized there were bugs and warnings being thrown, so just added in some additional fixes without really intending to optimize the code any further)

Comment (samukweku) [reply janitor/functions/truncate_datetime.py]: Yea, apply or if there is a faster option.

Comment (thatlittleboy) [reply janitor/functions/truncate_datetime.py]: Ok, I've modified it to use applymap so it's much faster now (5s -> 80ms on a df with 10k timestamps), and skip the df copying if a no-op is expected (when there are no datetime columns).

Comment (samukweku) [reply janitor/functions/truncate_datetime.py]: Sweeeeeet

**Comment by ericmjl** [comment]
> - not iterating over rows, and instead using .applymap on relevant columns only yields about 60x improvement. (5+ seconds -> 80ms on a df with 10k timestamps).

😮! That is impressive, @thatlittleboy!

**Review by ericmjl — APPROVED** [review]
LGTM and I think we can merge. Thank you @thatlittleboy! @samukweku would you like to do the honors?

**Comment by thatlittleboy** [comment]
Sorry, hold on one sec before merging, just realized one of the test docstring wasnt updated. i'll Need to push a quick fix

EDIT: okays, done!

**Pull request merged**